src/repositories/request-repository.ts

```diff
@@ -145,4 +147,68 @@ export class RequestRepository extends Repository<Request> {
       })
       .getMany()
   }
+
+  /**
+   * Marks requests as READY if (in the order of precendence):
+   *  1. there are PROCESSING requests that need to be anchored and retried (the maximum anchoring delay has elapsed and the request hasn't been updated in a long time)
+   *  2. there are PENDING requests that need to be anchored (the maximum anchoring delay has elasped)
+   *  3. there are streamLimit streams needing an anchor (prioritizing PROCESSING requests that need to be retried, then PENDING requests)
+   * Returns the original requests that were marked as READY
+   */
+  public async findAndMarkReady(streamLimit: number): Promise<Request[]> {
+    const anchoringDeadline = new Date(Date.now() - MAX_ANCHORING_DELAY_MS)
+    const retryDeadline = new Date(Date.now() - PROCESSING_TIMEOUT)
+    const isolationLevel =
+      this.connection.options.type === 'sqlite' ? 'SERIALIZABLE' : 'REPEATABLE READ'
+
+    return this.connection.transaction(isolationLevel, async (transactionalEntityManager) => {
+      // retrieves up to streamLimit unique streams with their earliest request createdAt value.
+      // this will only return streams associated with requests that are PENDING, or PROCESSING and needs to be retried
+      const streamsToAnchor = await transactionalEntityManager
+        .getRepository(Request)
+        .createQueryBuilder('request')
+        .select(['request.streamId', 'request.createdAt'])
+        .where('request.status = :processingStatus AND request.updatedAt < :retryDeadline', {
+          processingStatus: RequestStatus.PROCESSING,
+          retryDeadline: DateUtils.mixedDateToUtcDatetimeString(retryDeadline),
```

Comment:
any idea why this is necessary here but hasn't been necessary in any of the other queries we run that use timestamps?

Reply:
this is where things get weird: I have this demo here https://github.com/ceramicnetwork/ceramic-anchor-service/blob/f1d17651dc1e266a4091fca2787a021dc503c657/src/repositories/__tests__/request-repository.test.ts#L414 that displays the different behaviours of the possible ways to do this date comparison. The only surefire way to retrieve a request by their date is by using this method. I have read sqlite does some conversion to the date based on timezones. I am not sure how to proceed HAHA.